File: backend/src/main/java/com/ecommerce/backend/service/ReviewService.java
package com.ecommerce.backend.service;

import com.ecommerce.backend.model.Product;
import com.ecommerce.backend.model.Review;
import com.ecommerce.backend.model.User;
import com.ecommerce.backend.repository.OrderRepository;
import com.ecommerce.backend.repository.ProductRepository;
import com.ecommerce.backend.repository.ReviewRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.util.List;

@Service
@Transactional
public class ReviewService {

    @Autowired
    private ReviewRepository reviewRepository;

    @Autowired
    private ProductRepository productRepository;

    @Autowired
    private OrderRepository orderRepository;

    public Review addReview(User user, Long productId, String comment, Integer rating) {
        Product product = productRepository.findById(productId)
                .orElseThrow(() -> new RuntimeException("Mahsulot topilmadi!"));

        if (rating == null || rating < 1 || rating > 5) {
            throw new RuntimeException("Baholash 1 dan 5 gacha bo'lishi kerak!");
        }

        // Check if user has purchased the product and order status is DELIVERED
        boolean hasPurchased = orderRepository.existsDeliveredOrderByUserAndProduct(user.getId(), productId);
        if (!hasPurchased) {
            throw new RuntimeException("Siz ushbu mahsulotni sotib olmagansiz yoki buyurtmangiz hali yetkazib berilmagan!");
        }

        // Check if user already reviewed this product
        boolean alreadyReviewed = reviewRepository.existsByUserIdAndProductId(user.getId(), productId);
        if (alreadyReviewed) {
            throw new RuntimeException("Siz ushbu mahsulotga allaqachon sharh qoldirgansiz!");
        }

        Review review = Review.builder()
                .product(product)
                .user(user)
                .comment(comment)
                .rating(rating)
                .build();

        return reviewRepository.save(review);
    }

    public List<Review> getReviewsByProductId(Long productId) {
        return reviewRepository.findByProductIdOrderByCreatedAtDesc(productId);
    }

    public boolean canUserReviewProduct(Long userId, Long productId) {
        if (userId == null || productId == null) {
            return false;
        }
        return orderRepository.existsDeliveredOrderByUserAndProduct(userId, productId)
                && !reviewRepository.existsByUserIdAndProductId(userId, productId);
    }

    public List<Long> getReviewedProductIds(Long userId) {
        if (userId == null) {
            return List.of();
        }
        return reviewRepository.findProductIdsByUserId(userId);
    }

    public List<Review> getAllReviews() {
        return reviewRepository.findAllByOrderByCreatedAtDesc();
    }

    public void deleteReview(Long id) {
        if (!reviewRepository.existsById(id)) {
            throw new RuntimeException("Sharh topilmadi!");
        }
        reviewRepository.deleteById(id);
    }

    public Review replyToReview(Long reviewId, User admin, String replyText) {
        Review review = reviewRepository.findById(reviewId)
                .orElseThrow(() -> new RuntimeException("Sharh topilmadi!"));

        if (replyText == null || replyText.trim().isEmpty()) {
            throw new RuntimeException("Javob matni bo'sh bo'lishi mumkin emas!");
        }

        review.setReplyText(replyText.trim());
        review.setReplier(admin);
        review.setReplyCreatedAt(LocalDateTime.now());

        return reviewRepository.save(review);
    }

    public Review deleteReply(Long reviewId) {
        Review review = reviewRepository.findById(reviewId)
                .orElseThrow(() -> new RuntimeException("Sharh topilmadi!"));

        review.setReplyText(null);
        review.setReplier(null);
        review.setReplyCreatedAt(null);

        return reviewRepository.save(review);
    }

    public List<Review> getReviewsByUserId(Long userId) {
        return reviewRepository.findByUserIdOrderByCreatedAtDesc(userId);
    }
}

File: frontend/src/app/services/review.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';

@Injectable({
  providedIn: 'root'
})
export class ReviewService {
  private apiUrl = 'http://localhost:8080/api/reviews';

  constructor(private http: HttpClient) {}

  getReviews(productId: number): Observable<any[]> {
    return this.http.get<any[]>(`${this.apiUrl}/product/${productId}`);
  }

  addReview(productId: number, comment: string, rating: number): Observable<any> {
    return this.http.post<any>(this.apiUrl, { productId, comment, rating });
  }

  canReview(productId: number): Observable<{ canReview: boolean }> {
    return this.http.get<{ canReview: boolean }>(`${this.apiUrl}/product/${productId}/can-review`);
  }

  getReviewedProductIds(): Observable<number[]> {
    return this.http.get<number[]>(`${this.apiUrl}/user/reviewed-products`);
  }

  getAllReviews(): Observable<any[]> {
    return this.http.get<any[]>(this.apiUrl);
  }

  deleteReview(id: number): Observable<any> {
    return this.http.delete<any>(`${this.apiUrl}/${id}`);
  }

  getMyReviews(): Observable<any[]> {
    return this.http.get<any[]>(`${this.apiUrl}/user`);
  }

  replyToReview(reviewId: number, replyText: string): Observable<any> {
    return this.http.post<any>(`${this.apiUrl}/${reviewId}/reply`, { replyText });
  }

  deleteReply(reviewId: number): Observable<any> {
    return this.http.delete<any>(`${this.apiUrl}/${reviewId}/reply`);
  }
}

